go/mysql/fakesqldb/server.go

```diff
@@ -419,12 +423,9 @@ func (db *DB) HandleQuery(c *mysql.Conn, query string, callback func(*sqltypes.R
 	for _, pat := range db.patternData {
 		if pat.expr.MatchString(query) {
 			userCallback, ok := db.queryPatternUserCallback[pat.expr]
+			db.mu.Unlock()
 			if ok {
-				// Since the user call back can be indefinitely stuck, we shouldn't hold the lock indefinitely.
-				// This is only test code, so no actual cause for concern.
-				db.mu.Unlock()
 				userCallback(query)
-				db.mu.Lock()
```

Comment:
We had to refactor this a bit - there's other places in this code that can end up blocking (e.g. the `BeforeFunc` call further above) and where we need to release the lock beforehand.

Instead of unlocking and relocking here, we changed the unlock calls to not be `defer`ed. We're not happy about that, but it seemed like the simplest way forward.

I think it would be nice to refactor the fake db server to clean some of this up in the future.